### PR TITLE
Reinstate optional login via custom browser tab

### PR DESCRIPTION
**Pull request by Tak**
#2767 
Maybe #2703
Maybe #2534
Maybe #2462

### CODE REVIEW

**Comment by connyduck** [comment]
Definitely a workaround for 2462, not sure about the other ones.

Works nicely, thanks.

Only thing I'm not sure about is the wording. The two options are "WebView" and "Browser" and not "Mastodon" and "Browser". It is always Mastodon (or whatever other platform).

Maybe "Login with Tusky" and "Login with Browser" would be better?

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -97,6 +97,7 @@
     <string name="action_more">More</string>
     <string name="action_compose">Compose</string>
     <string name="action_login">Log in with Mastodon</string>
+    <string name="action_browser_login">Login in Browser</string>
```
Comment (nikclayton) [inline app/src/main/res/values/strings.xml:100]:
```suggestion
    <string name="action_browser_login">Log in in Browser</string>
```

**Comment by nikclayton** [comment]
If I'm a user seeing this for the first time, what information do I need to know in order to make an informed choice?

Also, I'm a bit confused after reading https://github.com/tuskyapp/Tusky/issues/2767, which seems to make a strong case that the WebView option should go away entirely. The other issues don't make a case for keeping it either.

If it did go away the user wouldn't need to have a choice. Why is it still needed?

**Comment by connyduck** [comment]
Browser login does not work on all browsers, some users have very old or exotic ones. WebView login does not work for some 2-factor-auth schemes.

**Comment by nikclayton** [comment]
> Browser login does not work on all browsers, some users have very old or exotic ones. WebView login does not work for some 2-factor-auth schemes.

Thanks. @Tak please could that be added as motivating comments in the code.

I guess it's too late in the day to consider https://developer.android.com/training/id-auth/authenticate for this release. Is that on anyone's radar though?

I may have a solution that's workable for this though.

I'm going to guess that if the user has a browser on their device that is new enough to support the Custom Tabs protocol it's new enough to do OAuth properly.

https://developer.chrome.com/docs/android/custom-tabs/integration-guide/#how-can-i-check-whether-the-android-device-has-a-browser-that-supports-custom-tab explains how to check if such a browser is installed.

If it is, I suggest using it, and hiding the webview option if a custom-tabs supporting browser is installed.

This is probably the very common case.

You mentioned browser log in doesn't work on some old or exotic ones. Is there a list (even a partial one) of the package names? If one of those is on the device then the user could be warned appropriately and the webview option presented first.

A handful I checked:

- com.stoutner.privacybrowser.standard (https://github.com/tuskyapp/Tusky/issues/2040)
- Tor Browser, maybe (https://github.com/tuskyapp/Tusky/issues/1609)
- Default browser on Lineage OS (maybe, this is from 2018), https://github.com/tuskyapp/Tusky/issues/682
 
---

Separately, the FAQ should be updated to cover this.

**Comment by connyduck** [comment]
I'm wondering what we could do to better explain the two buttons for users. We could do tooltips on longpress but I don't think that pattern is common enough for users to find it by themselves.

Something like

Login with Tusky - works in most cases, no data leaked to other apps
Login with Browser - supports 2FA with Webauthn but requires compatible browser

**Comment by nikclayton** [comment]
> > I guess it's too late in the day to consider https://developer.android.com/training/id-auth/authenticate for this release. Is that on anyone's radar though?
> 
> Ugh we used to use this at work, it is a weird api. Don't see what the advantage is unless you want single sign on between multiple applications on one device.

As a user, it lets me:

- Easily sync my account info across multiple devices (I use Tusky on phone and tablet)
- Easily see the accounts my device knows about (in the settings), and modify them from one place
- (future/speculative) If other Mastodon Android apps did this I wouldn't need to re-enter my credentials in to each one

I'll think about this a bit, and possibly create a separate issue, to not derail this one any further.

> > ...
> 
> One problem is when we use custom tabs exclusively is that Google Play could block us again when we don't show the server rules during login...

This is https://support.google.com/googleplay/android-developer/answer/9876937 ?

At the moment the user flow for adding an account is:

1. Enter the instance name, click "Log in with Mastodon"
2. Enter the username and password

What if there's an additional step between those two, that displays the rules from /api/v{1,2}/instance, and requires the user to click through them to proceed?

Would that help?

**Comment by connyduck** [comment]
Looking at it now, I don't think we can do it like this.



The screen feels overwhelming now and I don't think the description makes it clear enough 😞 

I wish we wouldn't have that 🤬 rules requirement from Google Play.
Maybe we can make something like detecting available browsers work. If the available browser is not compatible we fall back to WebView login but display a warning that 2fa might not work. But another screen for the rules, really 😞?

**Comment by charlag** [comment]
Put a three dot option to log in via the other option (whatever is the primary). If we see that it didn't work out we can also suggest in the message.

**Comment by nikclayton** [comment]
> > What if there's an additional step between those two, that displays the rules from /api/v{1,2}/instance, and requires the user to click through them to proceed?
> 
> > Would that help?
> 
> yes but it slows signup down

Is that a big deal if (a) it's a one time thing, and (b) the risk of not doing it is the app being rejected from the Play store?

Is there an additional nuance I'm missing?

**Comment by connyduck** [comment]
> Put a three dot option to log in via the other option (whatever is the primary). If we see that it didn't work out we can also suggest in the message.

I like that idea

> Is that a big deal if (a) it's a one time thing, and (b) the risk of not doing it is the app being rejected from the Play store?

It is not that big of a deal, but I think there are better options

> Edit to add: The only argument I can think to make to Google is that Tusky is in the same boat as an IMAP mail client, and they presumably don't require those to show any rules before the user can log in to their account.

Yeah I tried that, it was like they didn't even read my message. Here is my rant from back then: https://chaos.social/@ConnyDuck/109004198955056199

**Comment by Tak** [comment]
Implemented the suggestion from https://github.com/tuskyapp/Tusky/pull/3165#issuecomment-1384584096

**Comment by nikclayton** [comment]
As I understand it, the goal here is to pick the most appropriate initial web login approach, with a user-accessible fallback if necessary.

In addition, some users may have either (a) browsers that just don't work, (b) webview implementations that just don't work. If we can detect a user in either of those states we can prompt them to install another browser, and explain why.

So I think an improved UX flow would be:

**1**: Figure out the log in technology (enum of CUSTOM_TABS, BROWSER, or WEBVIEW)

a. Get the list of installed packages that support custom tabs (see https://developer.chrome.com/docs/android/custom-tabs/integration-guide/#how-can-i-check-whether-the-android-device-has-a-browser-that-supports-custom-tab for code that does this).

If that list is non-empty, then it's CUSTOM_TABS. Go to **2**

b. Get the complete list of installed browsers (the code at the previous link can do that too). Compare the list against either an allow list of browser we know do work, or a deny list of browsers we know don't (https://github.com/tuskyapp/Tusky/pull/3165#issuecomment-1381087194 has an initial deny list).

If the list of browsers, when filtered to either browsers we know work, or excluding browsers we know don't, is non-empty, then the log in technology is BROWSER. Go to **2**

c. Fall back to WEBVIEW.

**2**. Display the log in UI

a. If CUSTOM_TABS, display something like the current UI. Possibly with a "..." menu to allow the user to force switch to another type, but I'd be surprised if that's necessary.

[Extra points: Use the `warmup()` method to preload the browser process so that the UX after the user starts logging in is faster]

b. If BROWSER, display something like the current UI, but with an additional warning, something like "Some browsers do not work with normal Mastodon log in. If logging in does not work you will need to install an additional web browser. The Tusky team knows that both [Firefox] and [Chrome] work.

c. If WEBVIEW, display something like the current UI, but with an additional warning, something like "Logging in normally requires a web browser. The built-in one on your device may be too old. If logging in does not work you will need to install a web browser. The Tusky team knows that both [Firefox] and [Chrome] work

With the [Firefox] and [Chrome] entries launching intents that will install those apps (see https://developer.android.com/distribute/marketing-tools/linking-to-google-play -- I know those URLs are play.google.com in the intent, but if the user has F-Droid installed it will respond to those intents too, see https://gitlab.com/fdroid/fdroidclient/-/blob/master/app/src/main/java/org/fdroid/fdroid/views/main/MainActivity.java#L260 for confirmation).

## Alternatively

As above, but drop WEBVIEW support entirely, and require the user to install a working browser, with a link to the FAQ that explains the security tradeoff.

---

I think that solves the problems -- the common case will work without needing UI changes, and in the BROWSER and WEBVIEW cases the user is given the tools to fix the problem.

We can argue about what should be in the list of browsers to recommend (I know Firefox and Chrome work, I have not been able to find a definitive list of other Android browsers that support custom tabs)

**Comment by DavidEdwards** [comment]
Some additional points:

1. WebView has security implications #2767 ([google link](https://developers.googleblog.com/2021/06/upcoming-security-changes-to-googles-oauth-2.0-authorization-endpoint.html)) — it should ideally not be an option, or as a last resort.
2. Browser can result in a failure in the case parental controls are active. I had an issue with Amazon Tablets being parental locked and users not being able to login with SSO in a Browser.
3. Chrome Tabs do not (if I remember correctly) have access to the saved logins of the user (especially if they don't use Chrome), so that cannot be streamlined.

I have had success in the case of SSO, using the following strategy:

1. Chrome Custom Tabs as the only displayed login strategy.
2. The user can still click to open their browser if they want to, allowing them mobility to use saved logins.

**Comment by connyduck** [comment]
Perfect is the enemy of good - I will merge this now so we have a solution for users affected by the SSO problem.
Please feel free to send PRs to improve on this.

**Pull request merged**